auth.go

```diff
@@ -0,0 +1,218 @@
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"os"
+
+	"github.com/mendersoftware/log"
+	"github.com/pkg/errors"
+)
+
+type AuthCode string
+
+type AuthReqData struct {
+	IdData      string `json:"id_data"`
+	TenantToken string `json:"tenant_token"`
+	Pubkey      string `json:"pubkey"`
+	SeqNumber   uint64 `json:"seq_no"`
+}
+
+type AuthRequest struct {
+	// request message data
+	Data []byte
+	// authorization code
+	Code AuthCode
+	// request signature
+	Signature []byte
+}
+
+// handler for authorization message data
+type AuthDataMessenger interface {
+	// build authorization request data, returns auth request and authorization
+	// code for the request
+	MakeAuthRequest() (*AuthRequest, error)
+	// receive authoiriation data response, returns error if response is invalid
+	RecvAuthResponse([]byte) error
+}
+
+type AuthManager interface {
+	// returns true if authorization data is current and valid
+	IsValid() bool
+	// return device's authoirization code
+	AuthCode() (AuthCode, error)
+	// check if device key is available
+	HasKey() bool
+	// generate device key (will overwrite an already existing key)
+	GenerateKey() error
+
+	AuthDataMessenger
+}
+
+const (
+	authTokenName       = "authtoken"
+	authTenantTokenName = "authtentoken"
+	authSeqName         = "authseq"
+
+	noAuthCode = AuthCode("")
+)
+
+type MenderAuthManager struct {
+	store    Store
+	keystore *Keystore
+	keyname  string
+	idsrc    IdentityDataGetter
+	seqnum   SeqnumGetter
+}
+
+func NewAuthManager(store Store, keyname string, idsrc IdentityDataGetter) AuthManager {
+	ks := NewKeystore(store)
+	if ks == nil {
+		return nil
+	}
+
+	mgr := &MenderAuthManager{
+		store:    store,
+		keystore: ks,
+		keyname:  keyname,
+		idsrc:    idsrc,
+		seqnum:   NewFileSeqnum(authSeqName, store),
+	}
+
+	if err := ks.Load(keyname); err != nil && IsNoKeys(err) == false {
+		log.Errorf("failed to load device keys: %s", err)
+		return nil
+	}
+
+	return mgr
+}
+
+func (m *MenderAuthManager) IsValid() bool {
+	adata, err := m.AuthCode()
+	if err != nil {
+		return false
+	}
+
+	if adata == noAuthCode {
+		return false
+	}
+
+	// TODO check if JWT is valid?
+
+	return true
+}
+
+func (m *MenderAuthManager) MakeAuthRequest() (*AuthRequest, error) {
+
+	var err error
+	authd := AuthReqData{}
+
+	idata, err := m.idsrc.Get()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain identity data")
+	}
+
+	authd.IdData = idata
+
+	// fill device public key
+	authd.Pubkey = m.keystore.PublicPEM()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain device public key")
+	}
+
+	tentok, err := m.store.ReadAll(authTenantTokenName)
```

Comment:
I am usually not complaining about this for local variables, but here I don't know why I always can see scout camp (and tent OK).

FYI
```
Finally, the convention in Go is to use MixedCaps or mixedCaps rather than underscores to write multiword names.
```